services/vision/obstacledepth/obstacle_depth.go

```diff
@@ -0,0 +1,376 @@
+		intrinsics: conf.intrinsics, returnPCDs: conf.ReturnPCDs, k: conf.K,
+	}
+	segmenter := myObsDep.buildObsDepthWithIntrinsics() // does the thing
+	return svision.NewService(name, r, nil, nil, nil, segmenter)
+}
+
+// buildObsDepthNoIntrinsics will return the shortest depth in the depth map as a Geometry point.
+func buildObsDepthNoIntrinsics() segmentation.Segmenter {
+	return func(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+		depthStream, err := src.Stream(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get stream from %s", src)
+		}
+		pic, release, err := depthStream.Next(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get image from stream %s", depthStream)
+		}
+		defer release()
+		dm, err := rimage.ConvertImageToDepthMap(ctx, pic)
+		if err != nil {
+			return nil, errors.New("could not convert image to depth map")
+		}
+		min, _ := dm.MinMax()
+
+		pt := spatialmath.NewPoint(r3.Vector{X: 0, Y: 0, Z: float64(min)}, "obstacle")
+		toReturn := make([]*vision.Object, 1)
+		toReturn[0] = &vision.Object{Geometry: pt}
+
+		return toReturn, nil
+	}
+}
+
+// buildObsDepthWithIntrinsics will use the methodology in Manduchi et al. to find obstacle points
+// before clustering and projecting those points into 3D obstacles.
+func (o *obsDepth) buildObsDepthWithIntrinsics() segmentation.Segmenter {
+	return func(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+		// Check if we have intrinsics here or in the camera properties. If not, don't even try
+		props, err := src.Properties(ctx)
+		if err != nil {
+			return nil, err
+		}
+		if props.IntrinsicParams == nil && o.intrinsics == nil {
+			return nil, errors.New("tried to build obstacles depth with intrinsics but no instrinsics found")
+		} else if props.IntrinsicParams != nil {
+			o.intrinsics = props.IntrinsicParams
+		}
+
+		depthStream, err := src.Stream(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get stream from %s", src)
+		}
+		pic, release, err := depthStream.Next(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get image from stream %s", depthStream)
+		}
+		defer release()
+		dm, err := rimage.ConvertImageToDepthMap(ctx, pic)
+		if err != nil {
+			return nil, errors.New("could not convert image to depth map")
+		}
+		o.dm = dm
+		o.makePointList(sampleN)
+		runtime.GOMAXPROCS(maxThreads)
+
+		// Use some goroutines to determine if each depth pixel is an obstacle
+		doneCh := make(chan bool, len(o.ptChunks))
+		var wg sync.WaitGroup
+		for i, chunk := range o.ptChunks {
+			wg.Add(1)
+			go func(i int, chunk []obsPoint) {
+				defer wg.Done()
+				for j, p := range chunk {
+					o.ptChunks[i][j].isObstacle = o.isObstaclePoint(p.point)
+				}
+				doneCh <- true
+			}(i, chunk)
+		}
+		go func() {
+			wg.Wait()
+			close(doneCh)
+		}()
+		count := 0
+		for c := range doneCh {
+			if c {
+				count++
+			}
+		}
```

Comment:
I don't think I understand the purpose of lines 211-220. Do we want to wait for all the goroutines to finish? If so, why not just call wg.Wait()?

Reply:
Done lol